storage/allocator.go

```diff
@@ -13,75 +13,289 @@
-	*proto.StoreDescriptor, error) {
+// getUsedNodes returns a set of node IDs which are already being used
+// to store replicas.
+func getUsedNodes(existing []proto.Replica) map[proto.NodeID]struct{} {
 	// Get a set of current nodes -- we never want to allocate on an existing node.
-	usedNodes := make(map[proto.NodeID]struct{})
-	for _, replica := range existingReplicas {
+	usedNodes := map[proto.NodeID]struct{}{}
+	for _, replica := range existing {
 		usedNodes[replica.NodeID] = struct{}{}
 	}
+	return usedNodes
+}
+
+// storeDescFromGossip retrieves a StoreDescriptor from the specified
+// capacity gossip key. Returns an error if the gossip doesn't exist
+// or is not a StoreDescriptor.
+func storeDescFromGossip(key string, g *gossip.Gossip) (*proto.StoreDescriptor, error) {
+	info, err := g.GetInfo(key)
 
-	stores, err := a.storeFinder(required)
 	if err != nil {
 		return nil, err
 	}
+	storeDesc, ok := info.(proto.StoreDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("gossiped info is not a StoreDescriptor: %+v", info)
+	}
+	return &storeDesc, nil
+}
+
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, _ bool) {
+	a.Lock()
+	defer a.Unlock()
+
+	// Clear the cached store lists on new gossip.
+	a.storeLists = nil
+	if a.capacityKeys == nil {
+		a.capacityKeys = map[string]struct{}{}
+	}
+	a.capacityKeys[key] = struct{}{}
+}
 
-	// Randomly pick a node weighted by capacity.
-	var candidates []*proto.StoreDescriptor
-	var capacityTotal float64
+// AllocateTarget returns a suitable store for a new allocation with the
+// required attributes. Nodes already accommodating existing replicas
+// are ruled out as targets.
+func (a *allocator) AllocateTarget(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	stores, err := a.selectRandom(3, required, existing, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Choose the store with the highest percent capacity available.
+	var leastStore *proto.StoreDescriptor
 	for _, s := range stores {
-		if _, ok := usedNodes[s.Node.NodeID]; !ok {
-			candidates = append(candidates, s)
-			capacityTotal += s.Capacity.PercentAvail()
+		if leastStore == nil || s.Capacity.FractionUsed() < leastStore.Capacity.FractionUsed() {
+			leastStore = s
+		}
+	}
+	return leastStore, nil
+}
+
+// RebalanceTarget returns a suitable store for a rebalance target
+// with required attributes. Rebalance targets are selected at random
+// from amongst stores which fall within an acceptable multiple of
+// standard deviations from either the used or range count means.
+func (a *allocator) RebalanceTarget(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	filter := func(s *proto.StoreDescriptor, count, used *stat) bool {
```

Comment:
Feels right to me. It's so integral to how we're using `selectRandom` from this method. When I was trying to share this code with `shouldRebalance()`, I broke it out. But when the sharing didn't work, I ended up wanting to keep it local as part of this method.